### DISCUSSION TITLE
"spin deploy": Allow deployments without CI/CD

### DISCUSSION THREAD

**Discussion by jaydrogers** [serversideup/spin#61] [Ideas]
# 👉 Describe the problem
[comment]: <> (What are you trying to solve?)
- The current version of Spin requires using GitHub Actions or GitLab CI for deployment, adding more knowledge requirements to get started
- Many users getting started are very confused on how to run deployments

# 👥 Problem evidence & reach
[comment]: <> (How many people have this problem?)
Some users have expressed interest in removing this requirement: https://discord.com/channels/910287105714954251/1195096786776359064

# 🥰 Describe the "impact" on users?
[comment]: <> (How will this make people's lives better once it is solved?)
- New users can deploy a server on their own very easily

# 🏆 How to solve this problem
[comment]: <> (Describe the feature that you are proposing and how it will solve the problem)

### Solution Overview
We can reduce the knowledge requirements by allowing users to run `spin provision` (to configure and prepare their server for Docker), then `spin deploy` to locally deploy their image to their servers.

![Spin Diagram](https://github.com/serversideup/spin/assets/3174134/10ea289b-d3bb-4b69-af5c-b2b0a4f3420d)

This proposed method eliminates a few headaches:
- Users don't need to learn GitHub Actions or the deep inner workings on how a Docker Registry works
- This configuration is a lot easier to debug since its working with systems directly in the user's control

> [!NOTE]  
> This method is intended to be used for **very simple** set ups. For most environments, it will be recommended to use GitHub Actions or GitLab CI (or comparable) to deploy applications

### Deployment Process
1. Initialize Project (create files)
2. Configure (ensure Spin configs are set)
3. Build (ensure tag is set correctly, ensure env exists within application)
4. Transfer (use `docker image save`: https://nickjanetakis.com/blog/docker-tip-97-save-and-load-docker-images-from-a-zip-file)
5. Run `docker stack deploy` to initialize the transfer

# 💯 How do we validate the problem is solved?
[comment]: <> (Explain use cases on how we can measure the success of this implementation)
- [ ] Users can run `spin provision` (to configure their server) then `spin deploy` to deploy their stack

**Comment by jaydrogers** [comment]
Hey @danmatthews,

You had a really good question that I wanted to bring out on this thread:

<img width="1680" alt="image" src="https://github.com/serversideup/spin/assets/3174134/8ea4ba26-4890-485e-99b2-3ddacb6b0be6">

**To answer your question:**
If the user uses GitHub Actions to deploy, it does two major things:
1. Deployments can run without any user interaction
2. GitHub Packages is running as a registry, and GitHub Actions uses it in a way that reduces a lot of headache with the user configuring their own Docker Registry.

**Main Points:**
- 👉 If the user chooses to run `spin deploy`, it will require a Docker registry to be running that both the workstation and the server has access to
- 👉 I assume users will want a **private** registry, which even adds more complexity (that's why my original thought was to have the user just use GitHub Actions)

**Reply by danmatthews** [reply]
Thanks @jaydrogers will keep an 👁️  on this issue.

**Comment by jaydrogers** [comment]
@danmatthews: I discovered a way on how I might be able to do this without GitHub Actions or any other CI/CD system.

@simonj: I know you had a **hell of a time** getting GitHub Actions to work. The method proposed above would eliminate the need for learning GitHub Actions and dealing with all the limitations you were running into before.

I am interested in hearing your thoughts if you guys have any on my proposal above.

**Comment by jaydrogers** [comment]
# 🥳 Big update
This has been developed and is available with Spin 2.0+ 

https://serversideup.net/open-source/spin/docs/command-reference/deploy
